### DISCUSSION TITLE
Software components

### DISCUSSION THREAD

**Discussion by BrendaQiu** [baharsener/visualEyes#26] [Ideas]
Leave your ideas here for what components we should have in our software!

**Comment by baharsener** [comment]
bare bones description of a component: something that is a "thing that needs to happen", has clear inputs and outputs

**Comment by BrendaQiu** [comment]
File I/O: a function that reads EDF input and output the data in a unified format (pandas dataframe, numpy array/dict, etc.)

**Comment by BrendaQiu** [comment]
Data wrangling: have a consistent way to handle missing data fields, missing data points, etc. so functions down the line can handle different `.edf`.

**Reply by mckenziephagen** [reply]
Eyelink IO deals with a lot of the wrangling so we might not need to write this component.

**Comment by baharsener** [comment]
One component we discussed was a **function that defines an area of interest (AOI)**
**Input** would be screen coordinates (x and y range- let's think of how we can make it a flexible shape?) that define the area (could it take multiple units? Or would it have to be defined in pixels/cm/inches/visual angle etc?). 

**Output** could be (either/or/other options?): 1) A binary mask

**Reply by baharsener** [reply]
AOI function could potentially take an array that has the shape, center point of AOI (coordinates), size of AOI and a label of AOI

**Comment by baharsener** [comment]
We could have a separate component that just describe task variables. Some examples:
- your screen size (think about flexibility of units)
- viewing distance (we also want a function that calculates visual angle)
- how many AOI's you have?

**Reply by BrendaQiu** [reply]
- shape of AOI (circle, square, etc.)

**Reply by mckenziephagen** [reply]
What is the input/output for this?

**Reply by baharsener** [reply]
This is for people who know their visual angle:
input: visual angle (degree), screen size/resolution (pixels), viewing distance (mm)(e.g., 1 degree around my AOI)
output: size in pixel (number of pixels that correspond to the degree of visual angle under a certain viewing distance)

- this could be put in the AOI function.

**Comment by baharsener** [comment]


**Reply by baharsener** [reply]
this was wrong lol-- nvm

**Comment by lydiayqzhang** [comment]
**thresholding**:
input: rows of datapoints (or x, y), AOI coordinates
output: percentage of data within the AOI

**Reply by lydiayqzhang** [reply]
the size of the input data doesn't matter as long as we have the binary mask
give the user a percentage of data inside/outside the mask

**Comment by baharsener** [comment]
What are some ways some of these packages deal with blinks? Do we want to identify and remove them? Do we treat it as 'missing' or just a gap?

**Reply by baharsener** [reply]
Knowing # of missing data points could help

**Reply by baharsener** [reply]
this is prob pretty easy and we don't need to have it as a separate function

**Comment by mckenziephagen** [comment]
Plotting function: 
Take in data grouped by function (see https://github.com/baharsener/visualEyes/blob/main/dev/tech_review/baharDataeyelinkio_example.ipynb), output heatmap.

**Reply by mckenziephagen** [reply]
option: superimpose AOI onto plot

**Reply by baharsener** [reply]
also to keep in mind: we have to make sure you are doing your analysis based on 0,0 being the top-left corner of the screen and max values (e.g., 1279,1023) being the bottom right corner of the screen (with the position going up as you move right and down).

**Comment by lydiayqzhang** [comment]
check if df['discrete']['blinks'] fully overlap with the missing data in df['samples'] and what other things can lead to missing data??

**Comment by baharsener** [comment]
How I reset timestamp in R during my data cleaning:
So maybe we can try to do something similar (if people have trials rather than one run that is the whole file).
I basically set 'start trial' as the first timestamp for the unique trial, then subtract the 'start trial' time from each timestamp to get 'new time' so each trial starts at 0 and iterates by your sample rate (2 ms in my case)

for (trial in unique_trials) {
this_trial <- subd_sub_merged %>% filter(TRIAL_INDEX == trial)
start_trial <- this_trial$TIMESTAMP[1] #get first timestamp per trial
this_trial$new_time <- this_trial$TIMESTAMP - start_trial

I will work on making this into a function
